File: hospital/model/dao/PatientDAO.js
/**
 * DAO Interface: PatientDAO
 *
 * Defines the contract for all patient-related database operations.
 * Concrete implementations live in dao/impl/PatientDAOImpl.js.
 *
 * Controllers MUST only call these methods — never raw DB queries.
 */
export default class PatientDAO {
  /**
   * Fetch all patients.
   * @returns {Promise<Patient[]>}
   */
  async getAll()                   { throw new Error('PatientDAO.getAll() not implemented'); }

  /**
   * Fetch a single patient by primary key.
   * @param {number} id
   * @returns {Promise<Patient|null>}
   */
  async getById(id)                { throw new Error('PatientDAO.getById() not implemented'); }

  /**
   * Fetch a patient by userId (linked User record).
   * @param {number} userId
   * @returns {Promise<Patient|null>}
   */
  async getByUserId(userId)        { throw new Error('PatientDAO.getByUserId() not implemented'); }

  /**
   * Search patients by name or phone.
   * @param {string} query
   * @returns {Promise<Patient[]>}
   */
  async search(query)              { throw new Error('PatientDAO.search() not implemented'); }

  /**
   * Insert a new patient record.
   * @param {Patient} patient
   * @returns {Promise<Patient>} the created record with generated ID
   */
  async create(patient)            { throw new Error('PatientDAO.create() not implemented'); }

  /**
   * Update an existing patient.
   * @param {number}  id
   * @param {Partial<Patient>} data
   * @returns {Promise<Patient>}
   */
  async update(id, data)           { throw new Error('PatientDAO.update() not implemented'); }

  /**
   * Delete a patient record.
   * @param {number} id
   * @returns {Promise<boolean>}
   */
  async delete(id)                 { throw new Error('PatientDAO.delete() not implemented'); }

  /**
   * Count total patients.
   * @returns {Promise<number>}
   */
  async count()                    { throw new Error('PatientDAO.count() not implemented'); }
}

File: hospital-backend/src/main/java/com/hospital/dao/PatientDAO.java
package com.hospital.dao;

import com.hospital.model.Patient;
import java.util.List;

/**
 * PatientDAO
 *
 * Interface defining the contract for all patient-related database operations.
 * The concrete implementation lives in {@link com.hospital.dao.impl.PatientDAOImpl}.
 *
 * Rules:
 *   - No SQL, no JDBC, no business logic here — signatures only.
 *   - All methods throw {@link DAOException} (unchecked) on DB failure.
 *   - Servlets call this interface; they never instantiate DAOImpl directly
 *     (swap implementation without touching controllers).
 */
public interface PatientDAO {

    /**
     * Returns all patient records ordered by patient_id ascending.
     *
     * @return list of all patients; empty list if none exist
     * @throws DAOException on any DB error
     */
    List<Patient> getAll();

    /**
     * Returns the patient with the given primary key.
     *
     * @param id the patient_id
     * @return the matching {@link Patient}, or {@code null} if not found
     * @throws DAOException on any DB error
     */
    Patient getById(int id);

    /**
     * Returns the patient record linked to the given user account.
     * Used to resolve the logged-in user's patient profile from their session userId.
     *
     * @param userId the user_id foreign key
     * @return the matching {@link Patient}, or {@code null} if not found
     * @throws DAOException on any DB error
     */
    Patient getByUserId(int userId);

    /**
     * Inserts a new patient record and returns the generated patient_id.
     *
     * @param patient the patient to insert (patientId is ignored — DB auto-generates it)
     * @return the generated patient_id
     * @throws DAOException on any DB error or constraint violation
     */
    int insert(Patient patient);

    /**
     * Updates an existing patient record identified by {@code patient.getPatientId()}.
     *
     * @param patient the patient with updated field values
     * @return number of rows affected (1 on success, 0 if not found)
     * @throws DAOException on any DB error
     */
    int update(Patient patient);

    /**
     * Deletes the patient record with the given primary key.
     *
     * @param id the patient_id to delete
     * @return number of rows affected (1 on success, 0 if not found)
     * @throws DAOException on any DB error
     */
    int delete(int id);
}
